### PR TITLE
Implement pcap based USB replay

### PR DESCRIPTION
**Pull request by benzea**
The ioctl-tree based replay is mostly useful for protocols that are fundamentally state-less. With state-full protocols, the ioctl-tree may randomly jump around, cause confusing situations and gets the order wrong during recording.

The pcap based replay is also able to play-back control transfers adding this feature to umockdev, even if outside of the original ioctl-tree implementation.
    
Note that umockdev does not support creating a recording. Recordings should be done using a separate tool such as wireshark.
    
Closes: #82, #117

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -9,6 +9,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       fail-fast: false
+      env: EXTRA_PACKAGES=libpcap-dev
```
Comment (martinpitt) [inline .github/workflows/tests.yml]: Please add that directly to tests/run-alpine instead.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Any chance you could include a little pcap trace file and run the replay in a test, to ensure that this stays working, and actually works across OSes and architectures?

File: .github/workflows/tests.yml (hunk 1 above)
Comment (benzea) [reply .github/workflows/tests.yml]: Yeah, I realised that now, trying to fix that up currently :-)

**Comment by benzea** [comment]
OK, I added a test. I just recorded plugging in a keyboard and hit some keys. Then did the minimal ioctls to grab two packets.

Not ideal, should exercise a large chunk of the relevant code, even if many corner cases re not covered.

File: src/umockdev-pcap.vala [hunk 2]
```diff
@@ -0,0 +1,251 @@
+
```
Comment (martinpitt) [inline src/umockdev-pcap.vala:1]: Can you please add the standard copyright header, with your copyright?

File: tests/test-umockdev-vala.vala [hunk 3]
```diff
@@ -737,6 +817,8 @@ main (string[] args)
   Test.add_func ("/umockdev-testbed-vala/usbfs_ioctl_tree_override_default_device", t_usbfs_ioctl_tree_override_default_device);
   Test.add_func ("/umockdev-testbed-vala/usbfs_ioctl_tree_xz", t_usbfs_ioctl_tree_xz);
 
+  Test.add_func ("/umockdev-testbed-vala/usbfs_ioctl_pcap", t_usbfs_ioctl_pcap);
```
Comment (martinpitt) [inline tests/test-umockdev-vala.vala:820]: Please squash the unit test into the previous commit that adds the feature.

**Review by martinpitt — CHANGES_REQUESTED** [review]
This looks great, many thanks! Can you please still adjust README.md to mention that umockdev-run can load pcap files, and add an example how to generate them with wireshark and then load them?

Cheers!

File: README.rst [hunk 4]
```diff
@@ -152,6 +165,46 @@ Command line: Record and replay PtP/MTP USB devices
...
+
+    Bus 001 Device 004: ID 06cb:00bd Synaptics, Inc. Prometheus MIS Touch Fingerprint Reader
+
+- Dump the sysfs device and udev properties:
+
+  ::
+
+    $ umockdev-record /dev/bus/usb/001/004 > fingerprint.umockdev
+
+- Use `wireshark` to record the bus in question (in this case bus 0). By default
```
Comment (martinpitt) [inline README.rst:188]: Did you mean "bus 001" here, or is that another kind of "bus"?

**Comment by benzea** [comment]
Hmm, sec, I am getting errors from libusb. I think I need to change an error code somewhere.

**Comment by benzea** [comment]
OK, good now. I was returning ENOENT rather than EAGAIN if there was no reapable URB.

**Review by martinpitt — APPROVED** [review]
Looks great to me, thank you!
